src/OddJobs/Job.hs

```diff
@@ -453,8 +465,21 @@ jobMonitor = do
         liftIO $ Dir.removePathForcibly f
 
 -- | Ref: 'jobPoller'
-jobPollingSql :: TableName -> Query
-jobPollingSql tname = "update " <> tname <> " set status = ?, locked_at = ?, locked_by = ?, attempts=attempts+1 WHERE id in (select id from " <> tname <> " where (run_at<=? AND ((status in ?) OR (status = ? and locked_at<?))) ORDER BY run_at ASC LIMIT 1 FOR UPDATE) RETURNING id"
+jobPollingSql :: TableNames -> Query
+jobPollingSql tnames =
+  "update " <> tnJob tnames <>
+  "\n  set status = ?, locked_at = now(), locked_by = ?, attempts=attempts+1" <>
+  "\nwhere id in ( select id from " <> tnJob tnames <> " j_out" <>
+  "\n              where (run_at<=now() AND (status in ? OR (status = ? and locked_at < now() - ? * interval '1 second')))" <>
+  "\n                and case when resource_id is null then true" <>
+  "\n                         else ( select count(id) from " <> tnJob tnames <> " j_in" <>
+  "\n                                where j_in.resource_id = j_out.resource_id and j_in.status = ?" <>
+  "\n                                  and j_in.locked_at >= now() - ? * interval '1 second' )" <>
+  "\n                            < coalesce(( select resource_limit from " <> tnResource tnames <> " r" <>
+  "\n                                         where r.resource_id = j_out.resource_id ), ?)" <>
+  "\n                    end" <>
```

Comment:
This is the core of the feature.  You can see that in the case where there is no resource, the performance should not really be affected.  The sum and the resource limit are both supported by indices, so hopefully, they are also performant, but testing is needed.  

By doing this all within the query, there is no need for any additional "bookkeeping" (and no need for the transaction management issues that would bring). This is why I was adamant about doing this inside the library.  Of course all of this could be externalized, but, like transaction management, that could be rather disruptive.

Reply:
Let's discuss just this SQL query first. The other stuff is about naming, style nit-picks, etc.